src/language_server/vfs.rs

```diff
@@ -0,0 +1,153 @@
+use std::collections::HashMap;
+use std::io;
+use std::io::ErrorKind;
+use std::sync::RwLock;
+
+use super::document::Document;
+
+use reqwest::blocking::Client;
+use tower_lsp::lsp_types::Url;
+
+#[derive(Debug)]
+pub struct VFS {
+    overlays: RwLock<HashMap<Url, RwLock<Document>>>,
+
+    http_client: Client,
+}
+
+impl VFS {
+    pub fn new() -> io::Result<VFS> {
+        let client = Client::builder()
+            .gzip(true)
+            .build()
+            .map_err(categorise_reqwest_error)?;
+
+        Ok(VFS {
+            overlays: RwLock::new(HashMap::new()),
+            http_client: client,
+        })
+    }
+
+    pub fn get_document_contents(&self, uri: &Url) -> io::Result<String> {
+        {
+            let hm = self.overlays.read().unwrap();
+
+            if let Some(doc_lock) = hm.get(uri) {
+                let document = doc_lock.read().unwrap();
+                return Ok(document.contents().to_string());
+            };
+        }
+
+        let contents = self.read_file_from_uri(uri)?;
+        self.create_document(uri, &contents);
+        Ok(contents)
+    }
+
+    pub fn modify_document<F>(&self, uri: &Url, f: F) -> Option<()>
+    where
+        F: Fn(&mut Document),
+    {
+        let hm = self.overlays.read().unwrap();
+        match hm.get(uri) {
+            Some(doc_lock) => {
+                let document = &mut *doc_lock.write().unwrap();
+
+                f(document);
+                Some(())
+            }
+            None => None,
+        }
+    }
+
+    #[allow(dead_code)]
```

Comment:
Dead code?

Reply:
Yes.  This function is no longer used with the current refactoring of the main language_server file, but I was intending to use it as part of checking document versions in the didChange notification handling.

If we are adding the language-server without that improvement then this function can be removed along with its annotation.